executor/distsql.go

```diff
@@ -676,17 +705,33 @@ func (e *IndexLookUpExecutor) Close() error {
 		return nil
 	}
 
+	e.mu.Lock()
 	close(e.finished)
-	// Drain the resultCh and discard the result, in case that Next() doesn't fully
-	// consume the data, background worker still writing to resultCh and block forever.
-	for range e.resultCh {
-	}
-	e.idxWorkerWg.Wait()
-	e.tblWorkerWg.Wait()
 	e.finished = nil
 	e.workerStarted = false
+	if e.pendingIndexResults != nil {
+		e.pendingIndexResults.Close()
+	}
+	for _, worker := range e.tableWorkers {
+		worker.mu.Lock()
+		if worker.tableReader != nil {
+			worker.tableReader.Close()
+			worker.tableReader = nil
+		}
+		worker.mu.Unlock()
+	}
 	e.memTracker = nil
 	e.resultCurr = nil
+	resultCh, idxWorkerWg, tblWorkerWg := e.resultCh, e.idxWorkerWg, e.tblWorkerWg
+	e.mu.Unlock()
+	// Drain the resultCh and discard the result, in case that Next() doesn't fully
+	// consume the data, background worker still writing to resultCh and block forever.
+	go func() {
```

Comment:
If the executor do not wait for task to exit, I'm afraid there will be some data race or chunk reuse problems....